File: app/src/main/java/com/example/babajidemustapha/survey/features/takesurvey/activities/SurveyAction.java
package com.example.babajidemustapha.survey.features.takesurvey.activities;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.RectF;
import android.os.Bundle;
import android.text.style.ReplacementSpan;
import android.view.Menu;
import android.view.MenuItem;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.fragment.app.FragmentStatePagerAdapter;
import androidx.viewpager.widget.PagerAdapter;
import androidx.viewpager.widget.ViewPager;

import com.example.babajidemustapha.survey.R;
import com.example.babajidemustapha.survey.features.takesurvey.fragments.AnswerSurvey;
import com.example.babajidemustapha.survey.features.takesurvey.fragments.SurveyDescription;
import com.google.android.material.tabs.TabLayout;

import java.util.ArrayList;
import java.util.List;

public class SurveyAction extends AppCompatActivity {

    /**
     * The {@link PagerAdapter} that will provide
     * fragments for each of the sections. We use a
     * {@link FragmentPagerAdapter} derivative, which will keep every
     * loaded fragment in memory. If this becomes too memory intensive, it
     * may be best to switch to a
     * {@link FragmentStatePagerAdapter}.
     */
    private SectionsPagerAdapter mSectionsPagerAdapter;

    /**
     * The {@link ViewPager} that will host the section contents.
     */
    private ViewPager mViewPager;
    TabLayout tabLayout;
    int survey_id;
    boolean online;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_survey_action);
        Bundle bundle = getIntent().getExtras();
        online = bundle.getBoolean("online");
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        mSectionsPagerAdapter = new SectionsPagerAdapter(getSupportFragmentManager());

        mSectionsPagerAdapter.addFragment(new SurveyDescription(), "Description");
        mSectionsPagerAdapter.addFragment(new AnswerSurvey(), "Take Survey");
        mViewPager = findViewById(R.id.container);
        tabLayout = findViewById(R.id.tabs);
        if(!online) {
            survey_id = bundle.getInt("ID");

            // Create the adapter that will return a fragment for each of the three
            // primary sections of the activity.

            // mSectionsPagerAdapter.addFragment(new ResponseList(), "Responses");
            // Set up the ViewPager with the sections adapter.

            mViewPager.setAdapter(mSectionsPagerAdapter);

            tabLayout.setupWithViewPager(mViewPager);
//            setupTabs();
//            if(bundle.getBoolean("from_notification")){
//                mViewPager.setCurrentItem(2,true);
//            }
        }
        else{
            setTitle("External Survey");
            mViewPager.setAdapter(mSectionsPagerAdapter);
            tabLayout.setupWithViewPager(mViewPager);
        }

    }

//    public void setupTabs(){
//        for(int i = 0; i<tabLayout.getTabCount();i++){
//           // tabLayout.getTabAt(i).setCustomView(R.layout.survey_item_layout);
//            Log.e("tab text",tabLayout.getTabAt(i).getText().toString());
//            //Log.e("count",db.getResponsesCount(survey_id)+"");
//            switch (tabLayout.getTabAt(i).getText().toString().toLowerCase()){
//                case "responses":
//                    SpannableString customText = new SpannableString("RESPONSES "+db.getResponsesCount(survey_id));
////                    customText.setSpan(new BackgroundColorSpan(0xFFFFFFFF) , 10, customText.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
////                    customText.setSpan(new ForegroundColorSpan(0xFF000000) , 10, customText.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
//                    customText.setSpan(new RoundedBackgroundSpan(this) , 10, customText.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
//                    View view = getLayoutInflater().inflate(R.layout.tab_layout_with_badge,null);
//                    TextView textView = view.findViewById(R.id.tabText);
//                    TextView textView2 = view.findViewById(R.id.tabBadge);
//                    textView.setText("RESPONSES");
//                    textView2.setText(db.getResponsesCount(survey_id)+"");
////                    textView.setTextColor(new ColorStateList(new int[][]{
////                            {},{}
////                    },new int[]{}));
//                    tabLayout.getTabAt(i).setCustomView(view);
//                    break;
//                default:
//
//                    break;
//
//            }
//        }
//    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
       // getMenuInflater().inflate(R.menu.menu_survey_action, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    /**
     * A placeholder fragment containing a simple view.
     */

    /**
     * A {@link FragmentPagerAdapter} that returns a fragment corresponding to
     * one of the sections/tabs/pages.
     */
    public class SectionsPagerAdapter extends FragmentPagerAdapter {

        private List<Fragment> fragments = new ArrayList<>();
        private List<String> fragmentsTitle = new ArrayList<>();

        public SectionsPagerAdapter(FragmentManager fm) {
            super(fm);
        }

        @Override
        public Fragment getItem(int position) {
            // getItem is called to instantiate the fragment for the given page.
            // Return a PlaceholderFragment (defined as a static inner class below).
            return fragments.get(position);
        }

        @Override
        public int getCount() {
            // Show 3 total pages.
            return fragments.size();
        }

        public void addFragment(Fragment fragment, String title){
            fragments.add(fragment);
            fragmentsTitle.add(title);
        }
        @Override
        public CharSequence getPageTitle(int position) {
          return fragmentsTitle.get(position);
        }
    }
    public class RoundedBackgroundSpan extends ReplacementSpan {


        private int backgroundColor = 0;
        private int textColor = 0;
        private int PADDING = 5;

        public RoundedBackgroundSpan(Context context) {
            super();
            backgroundColor = 0xFFFFFFFF;
            textColor = 0xFF000000;
        }

        @Override
        public void draw(Canvas canvas, CharSequence text, int start, int end, float x, int top, int y, int bottom, Paint paint) {
            float CORNER_RADIUS = (bottom - top);
            float textWidth = x + measureText(paint, text, start, end);
          //  float CORNER_RADIUS = ( measureText(paint, text, start, end) + ( 2 * PADDING)) /2 ;
            RectF rect = new RectF(x - PADDING, top, textWidth + PADDING, bottom);
            paint.setColor(backgroundColor);
            canvas.drawRoundRect(rect, CORNER_RADIUS, CORNER_RADIUS, paint);
            paint.setColor(textColor);
            canvas.drawText(text, start, end, x, y, paint);
        }

        @Override
        public int getSize(Paint paint, CharSequence text, int start, int end, Paint.FontMetricsInt fm) {
            return Math.round(paint.measureText(text, start, end)) + PADDING;
        }

        private float measureText(Paint paint, CharSequence text, int start, int end) {
            return paint.measureText(text, start, end);
        }
    }
}

File: app/src/main/java/com/example/babajidemustapha/survey/features/takesurvey/activities/SurveyAction.kt
package com.example.babajidemustapha.survey.features.takesurvey.activities

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.fragment.app.FragmentStatePagerAdapter
import androidx.lifecycle.lifecycleScope
import androidx.viewpager.widget.PagerAdapter
import androidx.viewpager.widget.ViewPager
import com.example.babajidemustapha.survey.R
import com.example.babajidemustapha.survey.features.takesurvey.fragments.AnswerQuestionFragment
import com.example.babajidemustapha.survey.features.takesurvey.viewmodel.AnswerSurveyViewModel
import com.jide.surveyapp.model.QuestionType
import com.jide.surveyapp.model.ResponseDetail
import com.jide.surveyapp.model.ResponseHeader
import com.jide.surveyapp.model.relations.QuestionWithOptionsAndResponse
import com.jide.surveyapp.model.relations.ResponseHeaderWithDetails
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.android.synthetic.main.activity_survey_action.*
import kotlinx.coroutines.flow.collect
import java.util.*

@AndroidEntryPoint
class SurveyAction : AppCompatActivity(), AnswerQuestionFragment.OnQuestionSelectedInteractionListener {

    /**
     * The [PagerAdapter] that will provide
     * fragments for each of the sections. We use a
     * [FragmentPagerAdapter] derivative, which will keep every
     * loaded fragment in memory. If this becomes too memory intensive, it
     * may be best to switch to a
     * [FragmentStatePagerAdapter].
     */
    private var mSectionsPagerAdapter: SectionsPagerAdapter? = null

    private val viewModel by viewModels<AnswerSurveyViewModel>()

    //    TabLayout tabLayout;
    var survey_id: String = ""
    var online: Boolean = false
    var questions: List<QuestionWithOptionsAndResponse> = emptyList()
    var currentPosition = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_survey_action)
        val bundle = intent.extras
        online = bundle?.getBoolean("online") ?: false
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        if (!online) {
            survey_id = bundle!!.getString("ID", "")
            viewModel.getQuestions(survey_id)
            viewPager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
                override fun onPageScrollStateChanged(state: Int) {

                }

                override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {

                }

                override fun onPageSelected(position: Int) {
                    currentPosition = position
                    when (position) {
                        0 -> {
                            next.visibility = View.VISIBLE
                            previous.visibility = View.GONE
                        }
                        questions.size - 1 -> {
                            next.visibility = View.GONE
                            previous.visibility = View.VISIBLE
                        }
                        else -> {
                            next.visibility = View.VISIBLE
                            previous.visibility = View.VISIBLE
                        }
                    }
                }

            })
            viewPager.offscreenPageLimit = questions.size

            lifecycleScope.launchWhenStarted {
                viewModel.questionsFlow.collect {
                    this@SurveyAction.questions = it
                    mSectionsPagerAdapter = SectionsPagerAdapter(supportFragmentManager, this@SurveyAction.questions)
                    viewPager.adapter = mSectionsPagerAdapter
                }
            }

        } else {
            title = "External Survey"
            viewPager.adapter = mSectionsPagerAdapter
            //            tabLayout.setupWithViewPager(mViewPager);
        }
        next.setOnClickListener {
            viewPager.setCurrentItem(currentPosition + 1, true)
            // viewPager.currentItem
        }
        previous.setOnClickListener {
            viewPager.setCurrentItem(currentPosition - 1, true)
        }

    }

    private fun validate(): Boolean {
        for ((index, question) in questions.withIndex()) {
            when (question.question.questionType) {
                QuestionType.SHORT_TEXT, QuestionType.LONG_TEXT, QuestionType.SINGLE_OPTION, QuestionType.MULTIPLE_OPTION, QuestionType.REACTIONS -> {
                    if (question.responses.isEmpty() && question.question.mandatory) {
                        viewPager.setCurrentItem(index, false)
                        Toast.makeText(this, "One or more mandatory questions have not been answered.", Toast.LENGTH_LONG).show()
                        return false
                    }
                }
                QuestionType.IMAGES -> TODO()
                else -> return false
            }
        }
        return true
    }

    private fun storeResponse() {
        val responseDetails: List<ResponseDetail> = questions.flatMap { it.responses }
        viewModel.saveResponse(ResponseHeaderWithDetails(ResponseHeader(null, survey_id), responseDetails))
        Toast.makeText(this@SurveyAction, "Your response has been recorded", Toast.LENGTH_SHORT).show()
        finish()
    }

    private fun calculateProgress(): Float {
        var count = 0
        forLoop@ for (question in questions) {
            when (question.question.questionType) {
                QuestionType.SHORT_TEXT, QuestionType.LONG_TEXT, QuestionType.SINGLE_OPTION, QuestionType.MULTIPLE_OPTION, QuestionType.REACTIONS -> {
                    if (question.responses.isNullOrEmpty()) {
                        continue@forLoop
                    }
                }
                QuestionType.IMAGES -> TODO()
                else -> {
                }
            }
            count++
        }
        return count.toFloat() * 100 / questions.size
    }

    override fun updateProgress() {
        progressBar.progress = calculateProgress().toInt()
    }

    override fun getTotal(): Int {
        return questions.size
    }

    override fun onQuestionSelected(index: Int): QuestionWithOptionsAndResponse {
        return questions[index]
    }


    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.take_survey_menu, menu)

        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId


        return if (id == R.id.save_answers) {
            if (validate()) {
                storeResponse()
            }
            true
        } else super.onOptionsItemSelected(item)

    }

    /**
     * A placeholder fragment containing a simple view.
     */

    /**
     * A [FragmentPagerAdapter] that returns a fragment corresponding to
     * one of the sections/tabs/pages.
     */
    inner class SectionsPagerAdapter(fm: FragmentManager, val questions: List<QuestionWithOptionsAndResponse>) : FragmentStatePagerAdapter(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {

        override fun getItem(position: Int): Fragment {
            return AnswerQuestionFragment.newInstance(position)
        }

        override fun getCount(): Int {
            return questions.size
        }
    }
}
